Marlin/buzzer.h

```diff
@@ -20,11 +20,67 @@
  *
  */
 
-#ifndef BUZZER_H
-  #define BUZZER_H
+#ifndef __BUZZER_H__
+#define __BUZZER_H__
 
-  #if HAS_BUZZER
-    void buzz(long duration, uint16_t freq);
-  #endif
+#include "Marlin.h"
+#include "circularqueue.h"
 
-#endif //BUZZER_H
+#define TONE_QUEUE_LENGTH 8
+
+struct tone_t {
+  uint16_t duration;
+  uint16_t frequency;
+};
+
+class Buzzer {
```

Comment:
That's pretty much it. Singletons are classes meant to be instantiated only once. And no, we don't get super-formal (e.g., including an "Instance" class method) because we instantiate them with a single declaration. The main reason I've been favoring using classes lately is to get better encapsulation and to reduce the amount of globals. But we're still dealing with slow and modest AVR hardware, so wherever we can get `static` linkage, it is the best idea. It does produce slightly smaller and faster code, and we should take every opportunity where we can shave off 10 bytes or 2 cycles here or there.